### PR TITLE
[ALLUXIO-3361] add fuse integration test

### PR DESCRIPTION
**Pull request by LuQQiu**
https://alluxio.atlassian.net/browse/ALLUXIO-3361

### CODE REVIEW

**Comment by LuQQiu** [comment]
@calvinjia, I added the fuse integration test to test fuse mount point interactions with Alluxio filesystem. 

The POSIX commands included in this test is `cat, chmod, cp, dd, head, ls, mv, tail` and working on adding other commands like `du` and `chgrp`.

The followings are two issues I occur and willing to hear your advice.

(1) The fuse integration test mounts Alluxio to a local folder once each test. Would it be better to mount it once for the entire fuse integration test? I haven't found easy ways to accomplish this. The mount behavior seems not that heavy and slow in my mac computer (can be mounted in 1-2 sec).

(2) I can use 
```
mvn -DfailIfNoTests=false -Dtest=alluxio.client.fuse.FuseFileSystemIntegrationTest test -Dmaven.javadoc.skip -Dfindbugs.skip -Dcheckstyle.skip -Dlicense.skip -rf :alluxio-tests
```
or intellij to run the tests on my local mac and all tests build success.

Other than Mac, I also need to make sure the tests run as expected in linux environment. Do we have any better approach to run those integration tests on Linux environment (maybe autobots) ? 

PTAL, thanks a lot!

**Comment by LuQQiu** [comment]
@calvinjia Added the tests for `chown` and `chgrp`. Fixed the `chown userName fileName` issue in `AlluxioFuseFileSystem`. Linux will provide the main group name while group name is not provided but mac will not, so I add the gid check in `AlluxioFuseFileSystem.chown`. PTAL, thanks

File: integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java [hunk 1]
```diff
@@ -149,34 +149,48 @@ public int chown(String path, @uid_t long uid, @gid_t long gid) {
...
         }
         options.setOwner(userName);
-        LOG.info("Change owner and group of file {} to {}:{}", path, userName, groupName);
-      } else {
-        LOG.info("Change group of file {} to {}", path, groupName);
       }
 
-      mFileSystem.setAttribute(uri, options);
+      String groupName = "";
+      if (gid != -1 && gid != 4294967295L) {
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply integration/fuse/src/main/java/alluxio/fuse/AlluxioFuseFileSystem.java]: Changed to constants:

```
  /**
   * 4294967295 is unsigned long -1, -1 means that uid or gid is not set.
   * 4294967295 or -1 occurs when chown without user name or group name.
   * Please view https://github.com/SerCeMan/jnr-fuse/issues/67 for more details.
   */
  private static final long ID_NOT_SET_VALUE = -1;
  private static final long ID_NOT_SET_VALUE_UNSIGNED = 4294967295L;
```
Thanks

File: tests/src/test/java/alluxio/client/fuse/FuseFileSystemIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    mAlluxioRoot = "/";
+    mMountPoint = AlluxioTestDirectory
+        .createTemporaryDirectory("FuseMountPoint").getAbsolutePath();
+
+    AlluxioFuseOptions options = new AlluxioFuseOptions(mMountPoint,
+        mAlluxioRoot, false, new ArrayList<>());
+    mFuseFileSystem = new AlluxioFuseFileSystem(mFileSystem, options);
+    mFuseThread = new Thread(() -> mFuseFileSystem.mount(Paths.get(mMountPoint),
+        true, false, new String[]{"-odirect_io"}));
+    mFuseThread.start();
```
[the remark this answers is not in the document]
Comment (LuQQiu) [reply tests/src/test/java/alluxio/client/fuse/FuseFileSystemIntegrationTest.java]: Now I do the mounting in @BeforeTest and clean in @AfterTest. Thanks!

**Comment by LuQQiu** [comment]
@calvinjia  Thanks for your advice. I addressed your comments and tested in both local mac and Linux cluster. The only issue here is our PR Jenkins test seems to run in instances without Fuse installed so that all the FuseFileSystemIntegrationTests will be skipped. I will contact Rico on Monday to confirm that and see if we can install fuse in those PR test instances.

PTAL, thanks!

**Comment by calvinjia** [comment]
@LuQQiu Sorry I gave the wrong annotation, it should be junit's `@BeforeClass` and not `@BeforeTest`. Also, we should do a namespace clean up in `@After` and only do the unmount in `@AfterClass`.

**Comment by LuQQiu** [comment]
@calvinjia I changed to use `@BeforeClass` to setup mount, umount in `@AfterClass` and cleanup in `@After`. In addition, I modified the block size configuration and changed `dd` test to create smaller files so that the test can be finished faster.

For `du` command test, Alluxio and POSIX have different usages. Alluxio namespace will show the file size in du command. If `-s` is given, we will provide the aggregate file size of files inside a folder. However, when i test the `du` in Alluxio-Fuse. It will show the block size that mac estimate that the file/folder will use. For example, if a folder contains four files of size 0, `du -s /folder` in Alluxio will show 0 of this folder. `du /folder` will show 8 which is 4k.

I feel that `du` in Alluxio-Fuse can only show relative sizes and is not that useful. I would suggest not including `du` in our integration test now. What's your opinion?

Thanks!

**Comment by LuQQiu** [comment]
@calvinjia  In addition, I have contacted @aaudiber about installing fuse in jenkins OS PR build (add install commands in https://github.com/Alluxio/alluxio/blob/master/dev/jenkins/Dockerfile) so that the `FuseFileSystemIntegrationTest` can run instead of skip.

Now I am blocked by the network....unable to download the docker on my mac.....

**Comment by calvinjia** [comment]
@LuQQiu For the `du` behavior, is that consistent with Linux's `du`? It does not need to be consistent with Alluxio's `du`.

**Comment by LuQQiu** [comment]
@calvinjia yes, the Alluxio-Fuse `du` behavior is consistent with Linux `du`.

**Comment by calvinjia** [comment]
@LuQQiu In that case, I think our `du` behavior is fine. Alluxio's `du` is not a real `du`, just using the same name because it is doing something similar.

**Comment by LuQQiu** [comment]
@calvinjia Would we consider adding `du` to integration test? The `du` command results may vary depends on systems.

**Comment by calvinjia** [comment]
@LuQQiu `du`'s semantic should be consistent on local file systems, I thought the difference was with distributed storage like Alluxio? If it is not consistent then we should do a check to make sure the OS is what we expect (ie. `AssumeTrue`).

**Comment by LuQQiu** [comment]
@calvinjia sure, i will double check the `du` behavior in mac and linux systems.

**Comment by LuQQiu** [comment]
jenkins, test this please

**Comment by aaudiber** [comment]
hm, not sure why it isn't rebuilding

test this please

**Comment by aaudiber** [comment]
Jenkins, test this please

**Comment by LuQQiu** [comment]
@calvinjia I am working on the `fuse du` inconsistency behavior in Linux and testing to see if my changes can fix it. Will open another PR when finished. Thanks!

**Pull request merged**